# Homework 2 
**Ozan Yeşil** 090190325

# Question 1

For this question use the World Bank Data for Turkey for the following indicators. Use [wbgapi](https://pypi.org/project/wbgapi/) for getting the data.

* [Literacy rate, adult female (SE.ADT.LITR.FE.ZS)](https://data.worldbank.org/indicator/SE.ADT.LITR.FE.ZS)
* [Labor force, female (SL.TLF.TOTL.FE.ZS)](https://data.worldbank.org/indicator/SL.TLF.TOTL.FE.ZS)
* [Poverty headcount ratio at national poverty lines (SI.POV.NAHC)](https://data.worldbank.org/indicator/SI.POV.NAHC)
* [Current health expenditure per capita (SH.XPD.CHEX.PC.CD)](https://data.worldbank.org/indicator/SH.XPD.CHEX.PC.CD)
* [GDP per capita (NY.GDP.PCAP.CD)](https://data.worldbank.org/indicator/NY.GDP.PCAP.CD)
* [Mortality rate, under-5 (SH.DYN.MORT)](https://data.worldbank.org/indicator/SH.DYN.MORT)


Using the [statsmodels](https://www.statsmodels.org/stable/index.html) library write the best linear regression model using child mortality as the dependent variable while the rest are considered as independent variables. Pay particular attention to the fact that the order of the variables put into the model significantly impacts the performance of the model. Choose the best model by considering

* with the minimum number of variables and their interactions,
* with the optimal ordering of the independent variables and their interactions,
* $R^2$-score of the model,
* statistical significance of the model coefficients,
* ANOVA analysis of the model.


## Importing Libraries

In [269]:
import pandas as pd
import numpy as np
import wbgapi as wb
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
import yfinance as yf
import statsmodels.api as sm

from urllib.request import urlopen
import statsmodels.api as sm
from statsmodels.formula.api import ols


For this question use the World Bank Data for Turkey for the following indicators. Use [wbgapi](https://pypi.org/project/wbgapi/) for getting the data.

* [Literacy rate, adult female (SE.ADT.LITR.FE.ZS)](https://data.worldbank.org/indicator/SE.ADT.LITR.FE.ZS)
* [Labor force, female (SL.TLF.TOTL.FE.ZS)](https://data.worldbank.org/indicator/SL.TLF.TOTL.FE.ZS)
* [Poverty headcount ratio at national poverty lines (SI.POV.NAHC)](https://data.worldbank.org/indicator/SI.POV.NAHC)
* [Current health expenditure per capita (SH.XPD.CHEX.PC.CD)](https://data.worldbank.org/indicator/SH.XPD.CHEX.PC.CD)
* [GDP per capita (NY.GDP.PCAP.CD)](https://data.worldbank.org/indicator/NY.GDP.PCAP.CD)
* [Mortality rate, under-5 (SH.DYN.MORT)](https://data.worldbank.org/indicator/SH.DYN.MORT)

### Getting World Bank data

In [270]:
workSpace = {'literacy': {'link': 'SE.ADT.LITR.FE.ZS'},
             'Labor force': {'link': 'SL.TLF.TOTL.FE.ZS'},
             'Poverty': {'link': 'SI.POV.NAHC'},
             'Health Expanditure': {'link': 'SH.XPD.CHEX.PC.CD'},
             'GDP': {'link': 'NY.GDP.PCAP.CD'},
             'Mortality': {'link': 'SH.DYN.MORT'},
            }

country = 'TUR'

for x,y in workSpace.items():
    link = y['link']
    y.update({'data': wb.data.DataFrame(link,economy = country)})

In [271]:
workSpace.keys()

dict_keys(['literacy', 'Labor force', 'Poverty', 'Health Expanditure', 'GDP', 'Mortality'])

In [272]:
workSpace['literacy']['data']

YR1960  YR1961  YR1962  YR1963  YR1964  YR1965  YR1966  YR1967  \
economy                                                                   
TUR         NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

         YR1968  YR1969  ...     YR2012     YR2013     YR2014     YR2015  \
economy                  ...                                               
TUR         NaN     NaN  ...  91.604523  92.138077  92.401817  92.645813   

            YR2016     YR2017  YR2018     YR2019  YR2020  YR2021  
economy                                                           
TUR      93.563011  93.498268     NaN  94.424042     NaN     NaN  

[1 rows x 62 columns]

I created a function `transpose` which return the transpose of the pandas Series which makes easier to interpret.

In [273]:
def transpose(dfr):
    return np.transpose(dfr)

In [274]:
literacy = transpose(workSpace['literacy']['data'])
labor_force = transpose(workSpace['Labor force']['data'])
poverty = transpose(workSpace['Poverty']['data'])
health = transpose(workSpace['Health Expanditure']['data'])
gdp = transpose(workSpace['GDP']['data'])
mortality = transpose(workSpace['Mortality']['data'])

Using the [statsmodels](https://www.statsmodels.org/stable/index.html) library write the best linear regression model using child mortality as the dependent variable while the rest are considered as independent variables. Pay particular attention to the fact that the order of the variables put into the model significantly impacts the performance of the model. Choose the best model by considering

* with the minimum number of variables and their interactions,
* with the optimal ordering of the independent variables and their interactions,
* $R^2$-score of the model,
* statistical significance of the model coefficients,
* ANOVA analysis of the model.

In [275]:
data = pd.DataFrame()
data['lt'] = literacy['TUR']
data['lf'] = labor_force['TUR']
data['pv'] = poverty['TUR']
data['hlt'] = health['TUR']
data['gdp'] = gdp['TUR']
data['mrt'] = mortality['TUR']

In [276]:
model_q1 = ols(' mrt ~ hlt * gdp * lf  * pv ', data = data).fit()
print(model_q1.summary())

                            OLS Regression Results                            
Dep. Variable:                    mrt   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 3.525e+11
Date:                Sun, 06 Nov 2022   Prob (F-statistic):           1.32e-06
Time:                        20:45:31   Log-Likelihood:                 181.00
No. Observations:                  16   AIC:                            -332.0
Df Residuals:                       1   BIC:                            -320.4
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept     -1.691e+04      0.597  -2.83e+04

C:\ProgramData\Anaconda3\lib\site-packages\scipy\stats\stats.py:1603: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=16
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


In [277]:
sm.stats.anova_lm(model_q1)

df        sum_sq       mean_sq             F        PR(>F)
hlt            1.0  1.835535e+02  1.835535e+02  1.312076e+12  5.557771e-07
gdp            1.0  3.336832e+02  3.336832e+02  2.385232e+12  4.122065e-07
hlt:gdp        1.0  8.270153e+01  8.270153e+01  5.911664e+11  8.279903e-07
lf             1.0  8.756432e+01  8.756432e+01  6.259266e+11  8.046711e-07
hlt:lf         1.0  1.083558e+00  1.083558e+00  7.745478e+09  7.233625e-06
gdp:lf         1.0  8.536515e-03  8.536515e-03  6.102064e+07  8.149702e-05
hlt:gdp:lf     1.0  2.058336e-01  2.058336e-01  1.471338e+09  1.659678e-05
pv             1.0  5.188735e-02  5.188735e-02  3.709007e+08  3.305607e-05
hlt:pv         1.0  1.973917e-01  1.973917e-01  1.410993e+09  1.694797e-05
gdp:pv         1.0  3.386320e-01  3.386320e-01  2.420607e+09  1.293951e-05
hlt:gdp:pv     1.0  2.356891e-01  2.356891e-01  1.684751e+09  1.551002e-05
lf:pv          1.0  6.153754e-02  6.153754e-02  4.398821e+08  3.035372e-05
hlt:lf:pv      1.0  1.890636e-01  1.890636e-01  1.351463e+09  1.731721e-05
gdp:lf:pv      1.0  3.725756e-01  3.725756e-01  2.663242e+09  1.233601e-05
hlt:gdp:lf:pv  1.0  8.709497e-02  8.709497e-02  6.225716e+08  2.551441e-05
Residual       1.0  1.398955e-10  1.398955e-10           NaN           NaN

In [278]:
model_q1_final = ols('mrt ~ lf:pv + pv', data = data).fit()
print(model_q1_final.summary())

                            OLS Regression Results                            
Dep. Variable:                    mrt   R-squared:                       0.974
Model:                            OLS   Adj. R-squared:                  0.970
Method:                 Least Squares   F-statistic:                     245.3
Date:                Sun, 06 Nov 2022   Prob (F-statistic):           4.75e-11
Time:                        20:45:32   Log-Likelihood:                -23.564
No. Observations:                  16   AIC:                             53.13
Df Residuals:                      13   BIC:                             55.45
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -8.1671      2.292     -3.563      0.0

C:\ProgramData\Anaconda3\lib\site-packages\scipy\stats\stats.py:1603: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=16
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


In [279]:
sm.stats.anova_lm(model_q1_final)

df      sum_sq     mean_sq           F        PR(>F)
lf:pv      1.0  254.895431  254.895431  185.967219  4.444727e-09
pv         1.0  417.620532  417.620532  304.688588  2.099379e-10
Residual  13.0   17.818412    1.370647         NaN           NaN

# Question 2

For this question use Yahoo's Finance API for the following tickers:

* Gold futures (GC=F)
* Silver futures (SI=F)
* Copper futures (HG=F)
* Platinum futures (PL=F)

1. Write the best linear regression model that explains gold futures closing prices in terms of opening prices of gold, silver, copper, and platinum futures.
2. Repeat the same for silver, copper and platinum prices.
3. Compare the models you obtained in Steps 1 and 2. Which model is better? How do you decide? Explain.

For this question use Yahoo's Finance API for the following tickers:

* Gold futures (GC=F)
* Silver futures (SI=F)
* Copper futures (HG=F)
* Platinum futures (PL=F)

# Getting data with `yfinance`

In [280]:
gold_futures = yf.download("GC=F")
silver_futures = yf.download("SI=F")
copper_futures = yf.download("HG=F")
plat_futures = yf.download("PL=F")

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


In [281]:
gold_futures

Open         High          Low        Close  \
Date                                                                            
2000-08-30 00:00:00-04:00   273.899994   273.899994   273.899994   273.899994   
2000-08-31 00:00:00-04:00   274.799988   278.299988   274.799988   278.299988   
2000-09-01 00:00:00-04:00   277.000000   277.000000   277.000000   277.000000   
2000-09-05 00:00:00-04:00   275.799988   275.799988   275.799988   275.799988   
2000-09-06 00:00:00-04:00   274.200012   274.200012   274.200012   274.200012   
...                                ...          ...          ...          ...   
2022-10-31 00:00:00-04:00  1641.800049  1641.800049  1633.400024  1635.900024   
2022-11-01 00:00:00-04:00  1630.800049  1653.699951  1630.800049  1645.000000   
2022-11-02 00:00:00-04:00  1650.800049  1664.699951  1634.000000  1645.699951   
2022-11-03 00:00:00-04:00  1629.199951  1629.199951  1615.099976  1627.300049   
2022-11-04 00:00:00-04:00  1631.699951  1686.400024  1631.099976  1685.699951   

                             Adj Close  Volume  
Date                                            
2000-08-30 00:00:00-04:00   273.899994       0  
2000-08-31 00:00:00-04:00   278.299988       0  
2000-09-01 00:00:00-04:00   277.000000       0  
2000-09-05 00:00:00-04:00   275.799988       2  
2000-09-06 00:00:00-04:00   274.200012       0  
...                                ...     ...  
2022-10-31 00:00:00-04:00  1635.900024     522  
2022-11-01 00:00:00-04:00  1645.000000     985  
2022-11-02 00:00:00-04:00  1645.699951     612  
2022-11-03 00:00:00-04:00  1627.300049    1427  
2022-11-04 00:00:00-04:00  1685.699951  285985  

[5567 rows x 6 columns]

1. Write the best linear regression model that explains gold futures closing prices in terms of opening prices of gold, silver, copper, and platinum futures.

In [282]:
df_q2 = pd.DataFrame()
df_q2['gf'] = gold_futures['Open']
df_q2['sf'] = silver_futures['Open']
df_q2['cp'] = copper_futures['Open']
df_q2['pp'] = plat_futures['Open']
df_q2['y'] = gold_futures['Close']
df_q2

gf         sf      cp          pp  \
Date                                                                    
2000-08-30 00:00:00-04:00   273.899994   4.950000  0.8790  593.900024   
2000-08-31 00:00:00-04:00   274.799988   4.920000  0.8850  589.000000   
2000-09-01 00:00:00-04:00   277.000000   5.035000  0.8780  588.000000   
2000-09-05 00:00:00-04:00   275.799988   4.990000  0.8960  602.000000   
2000-09-06 00:00:00-04:00   274.200012   5.000000  0.9050  603.000000   
...                                ...        ...     ...         ...   
2022-10-31 00:00:00-04:00  1641.800049  19.200001  3.4095  939.400024   
2022-11-01 00:00:00-04:00  1630.800049  19.125000  3.4945  959.799988   
2022-11-02 00:00:00-04:00  1650.800049  19.780001  3.4985  960.200012   
2022-11-03 00:00:00-04:00  1629.199951  19.235001  3.4455  933.400024   
2022-11-04 00:00:00-04:00  1631.699951  19.500000  3.4255  922.000000   

                                     y  
Date                                    
2000-08-30 00:00:00-04:00   273.899994  
2000-08-31 00:00:00-04:00   278.299988  
2000-09-01 00:00:00-04:00   277.000000  
2000-09-05 00:00:00-04:00   275.799988  
2000-09-06 00:00:00-04:00   274.200012  
...                                ...  
2022-10-31 00:00:00-04:00  1635.900024  
2022-11-01 00:00:00-04:00  1645.000000  
2022-11-02 00:00:00-04:00  1645.699951  
2022-11-03 00:00:00-04:00  1627.300049  
2022-11-04 00:00:00-04:00  1685.699951  

[5567 rows x 5 columns]

In [283]:
model_df1 = ols('y ~  gf * sf  * cp * pp', data = df_q2).fit()
model_df1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                  1.000
Method:                 Least Squares   F-statistic:                 6.716e+05
Date:                Sun, 06 Nov 2022   Prob (F-statistic):               0.00
Time:                        20:46:13   Log-Likelihood:                -18717.
No. Observations:                4864   AIC:                         3.747e+04
Df Residuals:                    4848   BIC:                         3.757e+04
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      15.1554      7.212      2.101      0.036       1.016      29.295
gf              1.0126      0.016     64.410      0.000       0.982       1.043
sf             -5.8012      2.154     -2.694      0.007     -10.023      -1.579
gf:sf           0.0039      0.001      2.964      0.003       0.001       0.007
cp             -3.9090      5.812     -0.673      0.501     -15.304       7.486
gf:cp          -0.0041      0.006     -0.682      0.496      -0.016       0.008
sf:cp           1.6544      0.597      2.771      0.006       0.484       2.825
gf:sf:cp       -0.0011      0.000     -2.841      0.005      -0.002      -0.000
pp             -0.0115      0.010     -1.177      0.239      -0.031       0.008
gf:pp       -3.399e-05   1.62e-05     -2.099      0.036   -6.57e-05   -2.25e-06
sf:pp           0.0071      0.002      3.135      0.002       0.003       0.012
gf:sf:pp    -3.957e-06   1.24e-06     -3.198      0.001   -6.38e-06   -1.53e-06
cp:pp          -0.0004      0.004     -0.106      0.915      -0.008       0.007
gf:cp:pp     1.132e-05   5.07e-06      2.233      0.026    1.38e-06    2.13e-05
sf:cp:pp       -0.0017      0.001     -3.240      0.001      -0.003      -0.001
gf:sf:cp:pp  9.338e-07   3.05e-07      3.060      0.002    3.36e-07    1.53e-06
==============================================================================
Omnibus:                     1461.317   Durbin-Watson:                   2.061
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            29329.102
Skew:                          -0.934   Prob(JB):                         0.00
Kurtosis:                      14.884   Cond. No.                     7.15e+09
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.15e+09. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [284]:
sm.stats.anova_lm(model_df1)

df        sum_sq       mean_sq             F    PR(>F)
gf              1.0  1.301888e+09  1.301888e+09  1.007453e+07  0.000000
sf              1.0  1.012435e-02  1.012435e-02  7.834625e-05  0.992938
gf:sf           1.0  4.431645e+02  4.431645e+02  3.429385e+00  0.064107
cp              1.0  6.749374e+01  6.749374e+01  5.222937e-01  0.469900
gf:cp           1.0  1.114371e+03  1.114371e+03  8.623453e+00  0.003334
sf:cp           1.0  1.879696e+02  1.879696e+02  1.454584e+00  0.227853
gf:sf:cp        1.0  8.849994e+00  8.849994e+00  6.848481e-02  0.793567
pp              1.0  2.480256e+02  2.480256e+02  1.919321e+00  0.165995
gf:pp           1.0  2.892406e+02  2.892406e+02  2.238260e+00  0.134698
sf:pp           1.0  5.554887e+01  5.554887e+01  4.298594e-01  0.512089
gf:sf:pp        1.0  4.644538e+00  4.644538e+00  3.594130e-02  0.849645
cp:pp           1.0  6.167476e+01  6.167476e+01  4.772640e-01  0.489697
gf:cp:pp        1.0  3.828791e+02  3.828791e+02  2.962872e+00  0.085260
sf:cp:pp        1.0  1.653146e+02  1.653146e+02  1.279271e+00  0.258091
gf:sf:cp:pp     1.0  1.210348e+03  1.210348e+03  9.366155e+00  0.002222
Residual     4848.0  6.264860e+05  1.292257e+02           NaN       NaN

In [285]:
model_df2 = ols('y ~  gf:sf:cp:pp  + gf:sf ', data = df_q2).fit()
model_df2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.941
Model:                            OLS   Adj. R-squared:                  0.941
Method:                 Least Squares   F-statistic:                 3.860e+04
Date:                Sun, 06 Nov 2022   Prob (F-statistic):               0.00
Time:                        20:46:14   Log-Likelihood:                -30424.
No. Observations:                4864   AIC:                         6.085e+04
Df Residuals:                    4861   BIC:                         6.087e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept     367.9242      3.271    112.485      0.000     361.512     374.337
gf:sf:cp:pp -4.056e-06   4.05e-08   -100.083      0.000   -4.14e-06   -3.98e-06
gf:sf           0.0506      0.000    196.057      0.000       0.050       0.051
==============================================================================
Omnibus:                      359.614   Durbin-Watson:                   0.045
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              566.301
Skew:                          -0.579   Prob(JB):                    1.07e-123
Kurtosis:                       4.205   Cond. No.                     2.68e+08
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.68e+08. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [286]:
sm.stats.anova_lm(model_df2)

df        sum_sq       mean_sq             F  PR(>F)
gf:sf:cp:pp     1.0  6.151856e+08  6.151856e+08  38754.252742     0.0
gf:sf           1.0  6.101696e+08  6.101696e+08  38438.267340     0.0
Residual     4861.0  7.716358e+07  1.587401e+04           NaN     NaN

**2. Repeat the same for silver, copper and platinum prices.**


# For Silver

In [287]:
df_q2_silver = pd.DataFrame()
df_q2_silver['gf1'] = gold_futures['Open']
df_q2_silver['sf1'] = silver_futures['Open']
df_q2_silver['cp1'] = copper_futures['Open']
df_q2_silver['pp1'] = plat_futures['Open']
df_q2_silver['y1'] = silver_futures['Close']
df_q2_silver

gf1        sf1     cp1         pp1  \
Date                                                                    
2000-08-30 00:00:00-04:00   273.899994   4.950000  0.8790  593.900024   
2000-08-31 00:00:00-04:00   274.799988   4.920000  0.8850  589.000000   
2000-09-01 00:00:00-04:00   277.000000   5.035000  0.8780  588.000000   
2000-09-05 00:00:00-04:00   275.799988   4.990000  0.8960  602.000000   
2000-09-06 00:00:00-04:00   274.200012   5.000000  0.9050  603.000000   
...                                ...        ...     ...         ...   
2022-10-31 00:00:00-04:00  1641.800049  19.200001  3.4095  939.400024   
2022-11-01 00:00:00-04:00  1630.800049  19.125000  3.4945  959.799988   
2022-11-02 00:00:00-04:00  1650.800049  19.780001  3.4985  960.200012   
2022-11-03 00:00:00-04:00  1629.199951  19.235001  3.4455  933.400024   
2022-11-04 00:00:00-04:00  1631.699951  19.500000  3.4255  922.000000   

                                  y1  
Date                                  
2000-08-30 00:00:00-04:00   4.930000  
2000-08-31 00:00:00-04:00   5.003000  
2000-09-01 00:00:00-04:00   5.004000  
2000-09-05 00:00:00-04:00   4.998000  
2000-09-06 00:00:00-04:00   4.983000  
...                              ...  
2022-10-31 00:00:00-04:00  19.125000  
2022-11-01 00:00:00-04:00  19.673000  
2022-11-02 00:00:00-04:00  19.600000  
2022-11-03 00:00:00-04:00  19.436001  
2022-11-04 00:00:00-04:00  20.915001  

[5567 rows x 5 columns]

In [288]:
model_df1_silver = ols('y1 ~  gf1 * sf1  * cp1 * pp1', data = df_q2_silver).fit()
model_df1_silver.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     y1   R-squared:                       0.999
Model:                            OLS   Adj. R-squared:                  0.999
Method:                 Least Squares   F-statistic:                 2.444e+05
Date:                Sun, 06 Nov 2022   Prob (F-statistic):               0.00
Time:                        20:46:21   Log-Likelihood:                -1277.0
No. Observations:                4864   AIC:                             2586.
Df Residuals:                    4848   BIC:                             2690.
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          -0.3831      0.200     -1.916      0.055      -0.775       0.009
gf1                -0.0005      0.000     -1.199      0.231      -0.001       0.000
sf1                 1.0907      0.060     18.268      0.000       0.974       1.208
gf1:sf1         -4.187e-06   3.67e-05     -0.114      0.909   -7.61e-05    6.77e-05
cp1                 0.4244      0.161      2.634      0.008       0.108       0.740
gf1:cp1            -0.0002      0.000     -0.940      0.347      -0.000       0.000
sf1:cp1            -0.0334      0.017     -2.015      0.044      -0.066      -0.001
gf1:sf1:cp1      6.855e-06   1.04e-05      0.660      0.509   -1.35e-05    2.72e-05
pp1                 0.0004      0.000      1.294      0.196      -0.000       0.001
gf1:pp1          7.164e-07   4.49e-07      1.596      0.111   -1.64e-07     1.6e-06
sf1:pp1            -0.0001   6.31e-05     -1.877      0.061      -0.000    5.26e-06
gf1:sf1:pp1      1.312e-08   3.43e-08      0.383      0.702   -5.41e-08    8.04e-08
cp1:pp1            -0.0003      0.000     -3.313      0.001      -0.001      -0.000
gf1:cp1:pp1      1.354e-07   1.41e-07      0.963      0.336    -1.4e-07    4.11e-07
sf1:cp1:pp1      3.377e-05   1.49e-05      2.270      0.023     4.6e-06    6.29e-05
gf1:sf1:cp1:pp1 -8.565e-09   8.46e-09     -1.013      0.311   -2.51e-08    8.02e-09
==============================================================================
Omnibus:                     3369.395   Durbin-Watson:                   1.995
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           383991.267
Skew:                          -2.458   Prob(JB):                         0.00
Kurtosis:                      46.250   Cond. No.                     7.15e+09
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.15e+09. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [289]:
sm.stats.anova_lm(model_df1_silver)

df         sum_sq        mean_sq             F  \
gf1                 1.0  275715.317303  275715.317303  2.776315e+06   
sf1                 1.0   88326.487667   88326.487667  8.894036e+05   
gf1:sf1             1.0       0.609806       0.609806  6.140444e+00   
cp1                 1.0       0.165920       0.165920  1.670732e+00   
gf1:cp1             1.0       0.428206       0.428206  4.311825e+00   
sf1:cp1             1.0       0.540430       0.540430  5.441862e+00   
gf1:sf1:cp1         1.0       0.090240       0.090240  9.086671e-01   
pp1                 1.0       1.006160       1.006160  1.013153e+01   
gf1:pp1             1.0       1.286550       1.286550  1.295491e+01   
sf1:pp1             1.0       0.809860       0.809860  8.154885e+00   
gf1:sf1:pp1         1.0       0.056599       0.056599  5.699249e-01   
cp1:pp1             1.0       0.139212       0.139212  1.401793e+00   
gf1:cp1:pp1         1.0       2.659160       2.659160  2.677641e+01   
sf1:cp1:pp1         1.0       0.901145       0.901145  9.074081e+00   
gf1:sf1:cp1:pp1     1.0       0.101829       0.101829  1.025371e+00   
Residual         4848.0     481.453883       0.099310           NaN   

                       PR(>F)  
gf1              0.000000e+00  
sf1              0.000000e+00  
gf1:sf1          1.324622e-02  
cp1              1.962221e-01  
gf1:cp1          3.790090e-02  
sf1:cp1          1.970041e-02  
gf1:sf1:cp1      3.405159e-01  
pp1              1.466758e-03  
gf1:pp1          3.222709e-04  
sf1:pp1          4.312785e-03  
gf1:sf1:pp1      4.503254e-01  
cp1:pp1          2.364817e-01  
gf1:cp1:pp1      2.376232e-07  
sf1:cp1:pp1      2.605964e-03  
gf1:sf1:cp1:pp1  3.112989e-01  
Residual                  NaN

In [290]:
model_df1_silver_final = ols('y1 ~  gf1:sf1:cp1:pp1 + gf1:sf1:pp1 +gf1:pp1', data = df_q2_silver).fit()
model_df1_silver_final.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     y1   R-squared:                       0.962
Model:                            OLS   Adj. R-squared:                  0.962
Method:                 Least Squares   F-statistic:                 4.145e+04
Date:                Sun, 06 Nov 2022   Prob (F-statistic):               0.00
Time:                        20:46:22   Log-Likelihood:                -9422.3
No. Observations:                4864   AIC:                         1.885e+04
Df Residuals:                    4860   BIC:                         1.888e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           4.5882      0.059     77.328      0.000       4.472       4.705
gf1:sf1:cp1:pp1  2.475e-09   1.47e-09      1.678      0.093   -4.16e-10    5.37e-09
gf1:sf1:pp1       1.13e-07   7.63e-09     14.821      0.000    9.81e-08    1.28e-07
gf1:pp1          6.718e-06   1.02e-07     66.091      0.000    6.52e-06    6.92e-06
==============================================================================
Omnibus:                       36.586   Durbin-Watson:                   0.064
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               44.683
Skew:                           0.136   Prob(JB):                     1.98e-10
Kurtosis:                       3.383   Cond. No.                     3.78e+08
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.78e+08. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [291]:
sm.stats.anova_lm(model_df1_silver_final)

df         sum_sq        mean_sq              F  PR(>F)
gf1:sf1:cp1:pp1     1.0  315398.539059  315398.539059  111788.481602     0.0
gf1:sf1:pp1         1.0   23097.626017   23097.626017    8186.621754     0.0
gf1:pp1             1.0   12323.949366   12323.949366    4368.046825     0.0
Residual         4860.0   13711.939530       2.821387            NaN     NaN

# For Copper

In [292]:
df_q2_copper = pd.DataFrame()
df_q2_copper['gf2'] = gold_futures['Open']
df_q2_copper['sf2'] = silver_futures['Open']
df_q2_copper['cp2'] = copper_futures['Open']
df_q2_copper['pp2'] = plat_futures['Open']
df_q2_copper['y2'] = copper_futures['Close']
df_q2_copper

gf2        sf2     cp2         pp2      y2
Date                                                                         
2000-08-30 00:00:00-04:00   273.899994   4.950000  0.8790  593.900024  0.8850
2000-08-31 00:00:00-04:00   274.799988   4.920000  0.8850  589.000000  0.8850
2000-09-01 00:00:00-04:00   277.000000   5.035000  0.8780  588.000000  0.8890
2000-09-05 00:00:00-04:00   275.799988   4.990000  0.8960  602.000000  0.9060
2000-09-06 00:00:00-04:00   274.200012   5.000000  0.9050  603.000000  0.9015
...                                ...        ...     ...         ...     ...
2022-10-31 00:00:00-04:00  1641.800049  19.200001  3.4095  939.400024  3.4135
2022-11-01 00:00:00-04:00  1630.800049  19.125000  3.4945  959.799988  3.5095
2022-11-02 00:00:00-04:00  1650.800049  19.780001  3.4985  960.200012  3.5055
2022-11-03 00:00:00-04:00  1629.199951  19.235001  3.4455  933.400024  3.4565
2022-11-04 00:00:00-04:00  1631.699951  19.500000  3.4255  922.000000  3.7010

[5567 rows x 5 columns]

In [293]:
model_df1_copper = ols('y2 ~  gf2 * sf2  * cp2 * pp2', data = df_q2_copper).fit()
model_df1_copper.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     y2   R-squared:                       0.999
Model:                            OLS   Adj. R-squared:                  0.999
Method:                 Least Squares   F-statistic:                 2.730e+05
Date:                Sun, 06 Nov 2022   Prob (F-statistic):               0.00
Time:                        20:46:23   Log-Likelihood:                 8971.9
No. Observations:                4864   AIC:                        -1.791e+04
Df Residuals:                    4848   BIC:                        -1.781e+04
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           0.0100      0.024      0.412      0.680      -0.038       0.058
gf2              9.598e-05    5.3e-05      1.811      0.070   -7.91e-06       0.000
sf2                -0.0135      0.007     -1.857      0.063      -0.028       0.001
gf2:sf2          6.796e-06   4.46e-06      1.524      0.127   -1.94e-06    1.55e-05
cp2                 0.9916      0.020     50.611      0.000       0.953       1.030
gf2:cp2         -2.584e-05   2.02e-05     -1.279      0.201   -6.54e-05    1.38e-05
sf2:cp2             0.0036      0.002      1.776      0.076      -0.000       0.008
gf2:sf2:cp2     -1.565e-06   1.26e-06     -1.239      0.216   -4.04e-06    9.12e-07
pp2                8.9e-06    3.3e-05      0.270      0.788   -5.58e-05    7.36e-05
gf2:pp2         -1.516e-07   5.46e-08     -2.778      0.005   -2.59e-07   -4.46e-08
sf2:pp2          1.659e-05   7.67e-06      2.163      0.031    1.55e-06    3.16e-05
gf2:sf2:pp2     -6.508e-09   4.17e-09     -1.560      0.119   -1.47e-08    1.67e-09
cp2:pp2          -9.19e-07   1.27e-05     -0.073      0.942   -2.57e-05    2.39e-05
gf2:cp2:pp2      4.015e-08   1.71e-08      2.349      0.019    6.64e-09    7.37e-08
sf2:cp2:pp2      -4.07e-06   1.81e-06     -2.250      0.025   -7.62e-06   -5.23e-07
gf2:sf2:cp2:pp2  1.428e-09   1.03e-09      1.389      0.165   -5.88e-10    3.44e-09
==============================================================================
Omnibus:                      922.057   Durbin-Watson:                   2.133
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            18495.200
Skew:                           0.340   Prob(JB):                         0.00
Kurtosis:                      12.529   Cond. No.                     7.15e+09
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.15e+09. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [294]:
sm.stats.anova_lm(model_df1_copper)

df       sum_sq      mean_sq             F    PR(>F)
gf2                 1.0  4140.017827  4140.017827  2.819691e+06  0.000000
sf2                 1.0   432.284161   432.284161  2.944209e+05  0.000000
gf2:sf2             1.0   311.319240   311.319240  2.120339e+05  0.000000
cp2                 1.0  1128.755527  1128.755527  7.687749e+05  0.000000
gf2:cp2             1.0     0.000906     0.000906  6.173839e-01  0.432060
sf2:cp2             1.0     0.007191     0.007191  4.897484e+00  0.026942
gf2:sf2:cp2         1.0     0.000002     0.000002  1.575851e-03  0.968336
pp2                 1.0     0.003187     0.003187  2.170425e+00  0.140752
gf2:pp2             1.0     0.012807     0.012807  8.722679e+00  0.003158
sf2:pp2             1.0     0.000522     0.000522  3.556050e-01  0.550985
gf2:sf2:pp2         1.0     0.000125     0.000125  8.523974e-02  0.770331
cp2:pp2             1.0     0.000968     0.000968  6.593990e-01  0.416812
gf2:cp2:pp2         1.0     0.001015     0.001015  6.911259e-01  0.405823
sf2:cp2:pp2         1.0     0.007302     0.007302  4.973505e+00  0.025784
gf2:sf2:cp2:pp2     1.0     0.002831     0.002831  1.928070e+00  0.165034
Residual         4848.0     7.118087     0.001468           NaN       NaN

In [295]:
model_df1_copper_final = ols('y2 ~  gf2:sf2:cp2:pp2 + gf2:cp2:pp2 + gf2:cp2', data = df_q2_copper).fit()
model_df1_copper_final.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     y2   R-squared:                       0.906
Model:                            OLS   Adj. R-squared:                  0.906
Method:                 Least Squares   F-statistic:                 1.562e+04
Date:                Sun, 06 Nov 2022   Prob (F-statistic):               0.00
Time:                        20:46:24   Log-Likelihood:                -1668.4
No. Observations:                4864   AIC:                             3345.
Df Residuals:                    4860   BIC:                             3371.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           0.9753      0.010    101.512      0.000       0.956       0.994
gf2:sf2:cp2:pp2 -8.345e-09    1.6e-10    -52.244      0.000   -8.66e-09   -8.03e-09
gf2:cp2:pp2       4.55e-07   7.63e-09     59.636      0.000     4.4e-07     4.7e-07
gf2:cp2             0.0002   5.77e-06     29.566      0.000       0.000       0.000
==============================================================================
Omnibus:                     1596.335   Durbin-Watson:                   0.028
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4936.324
Skew:                           1.697   Prob(JB):                         0.00
Kurtosis:                       6.582   Cond. No.                     2.91e+08
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.91e+08. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [296]:
sm.stats.anova_lm(model_df1_copper_final)

df       sum_sq      mean_sq             F         PR(>F)
gf2:sf2:cp2:pp2     1.0  3295.606742  3295.606742  28321.704910   0.000000e+00
gf2:cp2:pp2         1.0  2056.680473  2056.680473  17674.650535   0.000000e+00
gf2:cp2             1.0   101.718897   101.718897    874.149378  8.184733e-177
Residual         4860.0   565.525586     0.116363           NaN            NaN

# For Platinium

In [297]:
df_q2_plat = pd.DataFrame()
df_q2_plat['gf3'] = gold_futures['Open']
df_q2_plat['sf3'] = silver_futures['Open']
df_q2_plat['cp3'] = copper_futures['Open']
df_q2_plat['pp3'] = plat_futures['Open']
df_q2_plat['y3'] = plat_futures['Close']
df_q2_plat

gf3        sf3     cp3         pp3  \
Date                                                                    
2000-08-30 00:00:00-04:00   273.899994   4.950000  0.8790  593.900024   
2000-08-31 00:00:00-04:00   274.799988   4.920000  0.8850  589.000000   
2000-09-01 00:00:00-04:00   277.000000   5.035000  0.8780  588.000000   
2000-09-05 00:00:00-04:00   275.799988   4.990000  0.8960  602.000000   
2000-09-06 00:00:00-04:00   274.200012   5.000000  0.9050  603.000000   
...                                ...        ...     ...         ...   
2022-10-31 00:00:00-04:00  1641.800049  19.200001  3.4095  939.400024   
2022-11-01 00:00:00-04:00  1630.800049  19.125000  3.4945  959.799988   
2022-11-02 00:00:00-04:00  1650.800049  19.780001  3.4985  960.200012   
2022-11-03 00:00:00-04:00  1629.199951  19.235001  3.4455  933.400024   
2022-11-04 00:00:00-04:00  1631.699951  19.500000  3.4255  922.000000   

                                   y3  
Date                                   
2000-08-30 00:00:00-04:00  591.400024  
2000-08-31 00:00:00-04:00  586.700012  
2000-09-01 00:00:00-04:00  595.299988  
2000-09-05 00:00:00-04:00  601.299988  
2000-09-06 00:00:00-04:00  611.099976  
...                               ...  
2022-10-31 00:00:00-04:00  939.400024  
2022-11-01 00:00:00-04:00  959.799988  
2022-11-02 00:00:00-04:00  960.200012  
2022-11-03 00:00:00-04:00  933.400024  
2022-11-04 00:00:00-04:00  967.700012  

[5567 rows x 5 columns]

In [298]:
model_df1_plat = ols('y3 ~  gf3 * sf3  * cp3 * pp3', data = df_q2_plat).fit()
model_df1_plat.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     y3   R-squared:                       0.999
Model:                            OLS   Adj. R-squared:                  0.999
Method:                 Least Squares   F-statistic:                 5.559e+05
Date:                Sun, 06 Nov 2022   Prob (F-statistic):               0.00
Time:                        20:46:26   Log-Likelihood:                -17495.
No. Observations:                4864   AIC:                         3.502e+04
Df Residuals:                    4848   BIC:                         3.513e+04
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept           5.4560      5.610      0.973      0.331      -5.542      16.454
gf3                 0.0321      0.012      2.624      0.009       0.008       0.056
sf3                -2.8786      1.675     -1.718      0.086      -6.163       0.406
gf3:sf3             0.0003      0.001      0.335      0.737      -0.002       0.002
cp3                -8.5618      4.521     -1.894      0.058     -17.425       0.302
gf3:cp3            -0.0069      0.005     -1.473      0.141      -0.016       0.002
sf3:cp3             1.3054      0.464      2.811      0.005       0.395       2.216
gf3:sf3:cp3        -0.0002      0.000     -0.745      0.456      -0.001       0.000
pp3                 0.9996      0.008    131.171      0.000       0.985       1.015
gf3:pp3         -3.302e-05   1.26e-05     -2.622      0.009   -5.77e-05   -8.33e-06
sf3:pp3             0.0027      0.002      1.530      0.126      -0.001       0.006
gf3:sf3:pp3     -3.399e-07   9.63e-07     -0.353      0.724   -2.23e-06    1.55e-06
cp3:pp3             0.0040      0.003      1.382      0.167      -0.002       0.010
gf3:cp3:pp3      6.159e-06   3.94e-06      1.562      0.118   -1.57e-06    1.39e-05
sf3:cp3:pp3        -0.0009      0.000     -2.098      0.036      -0.002   -5.73e-05
gf3:sf3:cp3:pp3  1.196e-07   2.37e-07      0.504      0.614   -3.46e-07    5.85e-07
==============================================================================
Omnibus:                     2502.671   Durbin-Watson:                   1.913
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           347459.196
Skew:                          -1.432   Prob(JB):                         0.00
Kurtosis:                      44.307   Cond. No.                     7.15e+09
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.15e+09. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [299]:
sm.stats.anova_lm(model_df1_plat)

df        sum_sq       mean_sq             F    PR(>F)
gf3                 1.0  1.636960e+08  1.636960e+08  2.093685e+06  0.000000
sf3                 1.0  2.832884e+08  2.832884e+08  3.623282e+06  0.000000
gf3:sf3             1.0  9.575605e+07  9.575605e+07  1.224728e+06  0.000000
cp3                 1.0  1.428077e+07  1.428077e+07  1.826523e+05  0.000000
gf3:cp3             1.0  1.110267e+06  1.110267e+06  1.420040e+04  0.000000
sf3:cp3             1.0  6.508193e+06  6.508193e+06  8.324034e+04  0.000000
gf3:sf3:cp3         1.0  3.681582e+06  3.681582e+06  4.708774e+04  0.000000
pp3                 1.0  8.360373e+07  8.360373e+07  1.069299e+06  0.000000
gf3:pp3             1.0  9.615291e+02  9.615291e+02  1.229804e+01  0.000458
sf3:pp3             1.0  3.098702e+02  3.098702e+02  3.963266e+00  0.046559
gf3:sf3:pp3         1.0  2.187311e+01  2.187311e+01  2.797589e-01  0.596883
cp3:pp3             1.0  1.645319e+01  1.645319e+01  2.104377e-01  0.646445
gf3:cp3:pp3         1.0  2.888760e+02  2.888760e+02  3.694748e+00  0.054643
sf3:cp3:pp3         1.0  1.031607e+03  1.031607e+03  1.319435e+01  0.000284
gf3:sf3:cp3:pp3     1.0  1.984218e+01  1.984218e+01  2.537831e-01  0.614447
Residual         4848.0  3.790437e+05  7.818557e+01           NaN       NaN

In [300]:
model_df1_plat_final = ols('y3 ~  gf3:sf3:cp3:pp3 + cp3:pp3 + gf3:sf3:pp3', data = df_q2_plat).fit()
model_df1_plat_final.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     y3   R-squared:                       0.926
Model:                            OLS   Adj. R-squared:                  0.926
Method:                 Least Squares   F-statistic:                 2.024e+04
Date:                Sun, 06 Nov 2022   Prob (F-statistic):               0.00
Time:                        20:46:27   Log-Likelihood:                -29286.
No. Observations:                4864   AIC:                         5.858e+04
Df Residuals:                    4860   BIC:                         5.861e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept         452.9156      3.173    142.755      0.000     446.696     459.136
gf3:sf3:cp3:pp3 -3.577e-06    7.8e-08    -45.857      0.000   -3.73e-06   -3.42e-06
cp3:pp3             0.1959      0.001    146.426      0.000       0.193       0.198
gf3:sf3:pp3      1.257e-05   3.16e-07     39.833      0.000     1.2e-05    1.32e-05
==============================================================================
Omnibus:                      513.498   Durbin-Watson:                   0.019
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              156.583
Skew:                          -0.126   Prob(JB):                     9.96e-35
Kurtosis:                       2.158   Cond. No.                     3.40e+08
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.4e+08. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [301]:
sm.stats.anova_lm(model_df1_copper_final)

df       sum_sq      mean_sq             F         PR(>F)
gf2:sf2:cp2:pp2     1.0  3295.606742  3295.606742  28321.704910   0.000000e+00
gf2:cp2:pp2         1.0  2056.680473  2056.680473  17674.650535   0.000000e+00
gf2:cp2             1.0   101.718897   101.718897    874.149378  8.184733e-177
Residual         4860.0   565.525586     0.116363           NaN            NaN

# Review

**3. Compare the models you obtained in Steps 1 and 2. Which model is better? How do you decide? Explain.**

In terms of R^2, model for Silver gave us the best results. Then gold, platinium and cooper, respectively. In each model confidence intervals were reliable, none of one had a different sign. Also, I used attributes that with  standard error and sum_sq values as lowest as possible.

# Question 3

1. Write a function that takes a ticker symbol and returns a pandas dataframe that for each day puts a 1 when the closing price is higher than the opening price, a 0 when the closing price is lower than the opening price.
2. Write a logistic regression that predicts the time series you obtain from Step 1 for gold futures against the opening prices of gold, silver, copper, and platinum prices.
3. Repeat the same for silver, copper, and platinum prices.
4. Compare the models you obtained from Steps 2 and 3. Decide which is the best model, and explain your reasoning.
5. Does any of the models provide a good fit? Explain.

In [302]:
from statsmodels.formula.api import logit
from statsmodels.api import Logit

**1. Write a function that takes a ticker symbol and returns a pandas dataframe that for each day puts a 1 when the closing price is higher than the opening price, a 0 when the closing price is lower than the opening price.**

I wrote a function named `ticker`. It takes the ticker symbol as the input and creates a list that gives 1 or 0 depending pn the comparison pf closing and opening price. Then, it adds that list as the new column to the dataframe and returns it. I named the new column as `Close > Open`.

In [303]:
def ticker(symbol):
    l = []
    yahoo_df = yf.download(str(symbol))
    for i in range(len(yahoo_df)):
        if(yahoo_df.iloc[i]['Close'] > yahoo_df.iloc[i]['Open']):
            l.append(1)
        else:
            l.append(0)
    yahoo_df['Close > Open'] = l
    return yahoo_df

In [304]:
ticker("GC=F")

[*********************100%***********************]  1 of 1 completed


Open         High          Low        Close  \
Date                                                                            
2000-08-30 00:00:00-04:00   273.899994   273.899994   273.899994   273.899994   
2000-08-31 00:00:00-04:00   274.799988   278.299988   274.799988   278.299988   
2000-09-01 00:00:00-04:00   277.000000   277.000000   277.000000   277.000000   
2000-09-05 00:00:00-04:00   275.799988   275.799988   275.799988   275.799988   
2000-09-06 00:00:00-04:00   274.200012   274.200012   274.200012   274.200012   
...                                ...          ...          ...          ...   
2022-10-31 00:00:00-04:00  1641.800049  1641.800049  1633.400024  1635.900024   
2022-11-01 00:00:00-04:00  1630.800049  1653.699951  1630.800049  1645.000000   
2022-11-02 00:00:00-04:00  1650.800049  1664.699951  1634.000000  1645.699951   
2022-11-03 00:00:00-04:00  1629.199951  1629.199951  1615.099976  1627.300049   
2022-11-04 00:00:00-04:00  1631.699951  1686.400024  1631.099976  1685.699951   

                             Adj Close  Volume  Close > Open  
Date                                                          
2000-08-30 00:00:00-04:00   273.899994       0             0  
2000-08-31 00:00:00-04:00   278.299988       0             1  
2000-09-01 00:00:00-04:00   277.000000       0             0  
2000-09-05 00:00:00-04:00   275.799988       2             0  
2000-09-06 00:00:00-04:00   274.200012       0             0  
...                                ...     ...           ...  
2022-10-31 00:00:00-04:00  1635.900024     522             0  
2022-11-01 00:00:00-04:00  1645.000000     985             1  
2022-11-02 00:00:00-04:00  1645.699951     612             0  
2022-11-03 00:00:00-04:00  1627.300049    1427             0  
2022-11-04 00:00:00-04:00  1685.699951  285985             1  

[5567 rows x 7 columns]

In [305]:
gc_q3 = ticker("GC=F")

[*********************100%***********************]  1 of 1 completed


In [306]:
si_q3 = ticker("SI=F")

[*********************100%***********************]  1 of 1 completed


In [307]:
hg_q3 = ticker("HG=F")

[*********************100%***********************]  1 of 1 completed


In [308]:
pl_q3 = ticker("PL=F")

[*********************100%***********************]  1 of 1 completed


2. Write a logistic regression that predicts the time series you obtain from Step 1 for gold futures against the opening prices of gold, silver, copper, and platinum prices.

In [309]:
df_q3 = pd.DataFrame()
df_q3['a'] = ticker("GC=F")['Open']
df_q3['b'] = ticker("SI=F")['Open']
df_q3['c'] = ticker("HG=F")['Open']
df_q3['d'] = ticker("PL=F")['Open']
df_q3['y'] = ticker("GC=F")['Close > Open']

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


In [310]:
df_q3

a          b       c           d  y
Date                                                                    
2000-08-30 00:00:00-04:00   273.899994   4.950000  0.8790  593.900024  0
2000-08-31 00:00:00-04:00   274.799988   4.920000  0.8850  589.000000  1
2000-09-01 00:00:00-04:00   277.000000   5.035000  0.8780  588.000000  0
2000-09-05 00:00:00-04:00   275.799988   4.990000  0.8960  602.000000  0
2000-09-06 00:00:00-04:00   274.200012   5.000000  0.9050  603.000000  0
...                                ...        ...     ...         ... ..
2022-10-31 00:00:00-04:00  1641.800049  19.200001  3.4095  939.400024  0
2022-11-01 00:00:00-04:00  1630.800049  19.125000  3.4945  959.799988  1
2022-11-02 00:00:00-04:00  1650.800049  19.780001  3.4985  960.200012  0
2022-11-03 00:00:00-04:00  1629.199951  19.235001  3.4455  933.400024  0
2022-11-04 00:00:00-04:00  1631.699951  19.500000  3.4255  922.000000  1

[5567 rows x 5 columns]

In [311]:
model_q3 = logit('y ~ a * b * c * d', data=df_q3).fit()
print(model_q3.summary())

Optimization terminated successfully.
         Current function value: 0.666797
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:                      y   No. Observations:                 4864
Model:                          Logit   Df Residuals:                     4848
Method:                           MLE   Df Model:                           15
Date:                Sun, 06 Nov 2022   Pseudo R-squ.:                 0.02224
Time:                        20:47:36   Log-Likelihood:                -3243.3
converged:                       True   LL-Null:                       -3317.1
Covariance Type:            nonrobust   LLR p-value:                 7.321e-24
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -2.6222      1.318     -1.990      0.047      -5.205      -0.040
a              0.0089      0.

In [312]:
model_q3 = logit('y ~ a + a:b:c+ a:d +b:d +a:c:d', data=df_q3).fit()
print(model_q3.summary())

Optimization terminated successfully.
         Current function value: 0.671680
         Iterations 4
                           Logit Regression Results                           
Dep. Variable:                      y   No. Observations:                 4864
Model:                          Logit   Df Residuals:                     4858
Method:                           MLE   Df Model:                            5
Date:                Sun, 06 Nov 2022   Pseudo R-squ.:                 0.01508
Time:                        20:47:36   Log-Likelihood:                -3267.1
converged:                       True   LL-Null:                       -3317.1
Covariance Type:            nonrobust   LLR p-value:                 5.123e-20
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.0799      0.104    -10.409      0.000      -1.283      -0.877
a              0.0016      0.

3. Repeat the same for silver, copper, and platinum prices.

# For Silver

In [313]:
df_q3_silver = pd.DataFrame()
df_q3_silver['y1'] = ticker("SI=F")['Close > Open']
df_q3_silver['a1'] = ticker("GC=F")['Open']
df_q3_silver['b1'] = ticker("SI=F")['Open']
df_q3_silver['c1'] = ticker("HG=F")['Open']
df_q3_silver['d1'] = ticker("PL=F")['Open']

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


In [314]:
model_q3_silver = logit('y1 ~ a1 * b1 * c1 * d1', data=df_q3_silver).fit()
model_q3_silver.summary()

Optimization terminated successfully.
         Current function value: 0.627000
         Iterations 5


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                     y1   No. Observations:                 4864
Model:                          Logit   Df Residuals:                     4848
Method:                           MLE   Df Model:                           15
Date:                Sun, 06 Nov 2022   Pseudo R-squ.:                 0.01985
Time:                        20:47:59   Log-Likelihood:                -3049.7
converged:                       True   LL-Null:                       -3111.5
Covariance Type:            nonrobust   LLR p-value:                 3.953e-19
===============================================================================
                  coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      -5.0508      1.380     -3.660      0.000      -7.755      -2.346
a1              0.0082      0.003      2.847      0.004       0.003       0.014
b1              0.2168      0.402      0.540      0.589      -0.570       1.004
a1:b1          -0.0002      0.000     -1.001      0.317      -0.001       0.000
c1              1.9580      1.101      1.778      0.075      -0.200       4.116
a1:c1          -0.0024      0.001     -2.202      0.028      -0.005      -0.000
b1:c1          -0.1014      0.113     -0.900      0.368      -0.322       0.119
a1:b1:c1     8.157e-05   6.99e-05      1.167      0.243   -5.55e-05       0.000
d1              0.0041      0.002      2.124      0.034       0.000       0.008
a1:d1       -9.594e-06   3.03e-06     -3.163      0.002   -1.55e-05   -3.65e-06
b1:d1          -0.0002      0.000     -0.358      0.720      -0.001       0.001
a1:b1:d1     2.796e-07   2.32e-07      1.207      0.227   -1.74e-07    7.33e-07
c1:d1          -0.0020      0.001     -2.851      0.004      -0.003      -0.001
a1:c1:d1     2.895e-06   9.32e-07      3.105      0.002    1.07e-06    4.72e-06
b1:c1:d1     9.064e-05      0.000      0.903      0.367      -0.000       0.000
a1:b1:c1:d1  -9.49e-08   5.68e-08     -1.672      0.095   -2.06e-07    1.63e-08
===============================================================================
"""

In [315]:
model_q3_silver = logit('y1 ~ a1 + d1 + a1:d1 +c1:d1 +a1:c1:d1 +a1:b1:c1:d1 ', data=df_q3_silver).fit()
model_q3_silver.summary()

Optimization terminated successfully.
         Current function value: 0.629959
         Iterations 5


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                     y1   No. Observations:                 4864
Model:                          Logit   Df Residuals:                     4857
Method:                           MLE   Df Model:                            6
Date:                Sun, 06 Nov 2022   Pseudo R-squ.:                 0.01522
Time:                        20:47:59   Log-Likelihood:                -3064.1
converged:                       True   LL-Null:                       -3111.5
Covariance Type:            nonrobust   LLR p-value:                 3.163e-18
===============================================================================
                  coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      -1.3408      0.457     -2.931      0.003      -2.237      -0.444
a1              0.0010      0.000      2.767      0.006       0.000       0.002
d1          -5.821e-05      0.001     -0.072      0.943      -0.002       0.002
a1:d1       -3.741e-07   5.59e-07     -0.669      0.503   -1.47e-06    7.22e-07
c1:d1           0.0002      0.000      1.438      0.150      -8e-05       0.001
a1:c1:d1    -2.872e-07   1.17e-07     -2.447      0.014   -5.17e-07   -5.72e-08
a1:b1:c1:d1  6.435e-09   1.36e-09      4.716      0.000    3.76e-09    9.11e-09
===============================================================================
"""

# For Copper

In [316]:
df_q3_copper = pd.DataFrame()
df_q3_copper['y2'] = ticker("HG=F")['Close > Open']
df_q3_copper['a2'] = ticker("GC=F")['Open']
df_q3_copper['b2'] = ticker("SI=F")['Open']
df_q3_copper['c2'] = ticker("HG=F")['Open']
df_q3_copper['d2'] = ticker("PL=F")['Open']

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


In [317]:
model_q3_copper = logit('y2 ~ a2 * b2 * c2 * d2', data=df_q3_copper).fit()
model_q3_copper.summary()

Optimization terminated successfully.
         Current function value: 0.687451
         Iterations 4


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                     y2   No. Observations:                 4864
Model:                          Logit   Df Residuals:                     4848
Method:                           MLE   Df Model:                           15
Date:                Sun, 06 Nov 2022   Pseudo R-squ.:                0.007939
Time:                        20:48:42   Log-Likelihood:                -3343.8
converged:                       True   LL-Null:                       -3370.5
Covariance Type:            nonrobust   LLR p-value:                 3.165e-06
===============================================================================
                  coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      -4.3367      1.286     -3.372      0.001      -6.857      -1.816
a2              0.0075      0.003      2.680      0.007       0.002       0.013
b2              0.0194      0.381      0.051      0.959      -0.728       0.767
a2:b2       -8.199e-05      0.000     -0.349      0.727      -0.001       0.000
c2              1.6064      1.032      1.556      0.120      -0.417       3.629
a2:c2          -0.0024      0.001     -2.215      0.027      -0.004      -0.000
b2:c2          -0.0274      0.106     -0.259      0.795      -0.234       0.180
a2:b2:c2     4.536e-05   6.66e-05      0.681      0.496   -8.51e-05       0.000
d2              0.0058      0.002      3.303      0.001       0.002       0.009
a2:d2       -1.093e-05   2.88e-06     -3.800      0.000   -1.66e-05   -5.29e-06
b2:d2        5.545e-05      0.000      0.138      0.891      -0.001       0.001
a2:b2:d2     1.675e-07    2.2e-07      0.762      0.446   -2.63e-07    5.98e-07
c2:d2          -0.0020      0.001     -3.005      0.003      -0.003      -0.001
a2:c2:d2     3.069e-06   9.01e-07      3.405      0.001     1.3e-06    4.84e-06
b2:c2:d2     1.967e-05   9.51e-05      0.207      0.836      -0.000       0.000
a2:b2:c2:d2 -6.591e-08   5.43e-08     -1.214      0.225   -1.72e-07    4.05e-08
===============================================================================
"""

In [318]:
model_q3_copper = logit('y2 ~ a2:b2:c2:d2  + a2:c2:d2 +a2:b2:d2', data=df_q3_copper).fit()
model_q3_copper.summary()

Optimization terminated successfully.
         Current function value: 0.692480
         Iterations 3


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                     y2   No. Observations:                 4864
Model:                          Logit   Df Residuals:                     4860
Method:                           MLE   Df Model:                            3
Date:                Sun, 06 Nov 2022   Pseudo R-squ.:               0.0006815
Time:                        20:48:42   Log-Likelihood:                -3368.2
converged:                       True   LL-Null:                       -3370.5
Covariance Type:            nonrobust   LLR p-value:                    0.2041
===============================================================================
                  coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      -0.0049      0.058     -0.084      0.933      -0.119       0.109
a2:b2:c2:d2 -1.698e-09    1.6e-09     -1.060      0.289   -4.84e-09    1.44e-09
a2:c2:d2    -4.013e-08    2.9e-08     -1.384      0.166    -9.7e-08    1.67e-08
a2:b2:d2     1.059e-08   6.54e-09      1.619      0.106   -2.23e-09    2.34e-08
===============================================================================
"""

# For Platinium

In [319]:
df_q3_plat = pd.DataFrame()
df_q3_plat['y3'] = ticker("PL=F")['Close > Open']
df_q3_plat['a3'] = ticker("GC=F")['Open']
df_q3_plat['b3'] = ticker("SI=F")['Open']
df_q3_plat['c3'] = ticker("HG=F")['Open']
df_q3_plat['d3'] = ticker("PL=F")['Open']

[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


In [320]:
model_q3_plat = logit('y3 ~ a3 * b3 * c3 * d3', data=df_q3_plat).fit()
model_q3_plat.summary()

Optimization terminated successfully.
         Current function value: 0.526528
         Iterations 6


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                     y3   No. Observations:                 4864
Model:                          Logit   Df Residuals:                     4848
Method:                           MLE   Df Model:                           15
Date:                Sun, 06 Nov 2022   Pseudo R-squ.:                 0.09106
Time:                        20:49:00   Log-Likelihood:                -2561.0
converged:                       True   LL-Null:                       -2817.6
Covariance Type:            nonrobust   LLR p-value:                9.383e-100
===============================================================================
                  coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept      -3.3737      1.503     -2.245      0.025      -6.320      -0.428
a3              0.0083      0.004      2.304      0.021       0.001       0.015
b3              0.4668      0.461      1.012      0.312      -0.437       1.371
a3:b3          -0.0007      0.000     -2.454      0.014      -0.001      -0.000
c3              0.0704      1.217      0.058      0.954      -2.315       2.456
a3:c3          -0.0031      0.001     -2.181      0.029      -0.006      -0.000
b3:c3          -0.0842      0.124     -0.681      0.496      -0.326       0.158
a3:b3:c3        0.0002   8.12e-05      2.890      0.004    7.55e-05       0.000
d3              0.0036      0.002      1.850      0.064      -0.000       0.007
a3:d3       -6.152e-06    3.6e-06     -1.710      0.087   -1.32e-05       9e-07
b3:d3          -0.0005      0.000     -1.001      0.317      -0.001       0.000
a3:b3:d3     5.685e-07   2.64e-07      2.151      0.031    5.05e-08    1.09e-06
c3:d3          -0.0001      0.001     -0.164      0.870      -0.002       0.002
a3:c3:d3     1.187e-06   1.18e-06      1.005      0.315   -1.13e-06     3.5e-06
b3:c3:d3        0.0001      0.000      1.061      0.288      -0.000       0.000
a3:b3:c3:d3 -1.737e-07    6.8e-08     -2.553      0.011   -3.07e-07   -4.03e-08
===============================================================================
"""

In [321]:
model_q3_plat = logit('y3 ~ a3 +a3:b3:d3 +a3:c3:d3 :b3 ', data=df_q3_plat).fit()
model_q3_plat.summary()

Optimization terminated successfully.
         Current function value: 0.537023
         Iterations 5


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                     y3   No. Observations:                 4864
Model:                          Logit   Df Residuals:                     4860
Method:                           MLE   Df Model:                            3
Date:                Sun, 06 Nov 2022   Pseudo R-squ.:                 0.07294
Time:                        20:49:00   Log-Likelihood:                -2612.1
converged:                       True   LL-Null:                       -2817.6
Covariance Type:            nonrobust   LLR p-value:                 8.985e-89
===============================================================================
                  coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept       0.3502      0.078      4.503      0.000       0.198       0.503
a3             -0.0013      0.000    -11.048      0.000      -0.002      -0.001
a3:b3:d3     -1.38e-08   1.02e-08     -1.357      0.175   -3.37e-08    6.13e-09
a3:c3:d3:b3   3.79e-09   2.29e-09      1.653      0.098   -7.04e-10    8.28e-09
===============================================================================
"""

4. Compare the models you obtained from Steps 2 and 3. Decide which is the best model, and explain your reasoning.

The highest Pseudo R-Squared among all is for **platinium**, silver, gold and copper, respectively. It is not possible to say that any model is good enough, actually we can say that the model are not useful with these accuracies. 

5. Does any of the models provide a good fit? Explain.

No, the highest Pseudo-R Squared value is 0.07 which is really weak.

# Question 4

For this question use the following [data](https://archive.ics.uci.edu/ml/datasets/credit+approval):


In [322]:
credit = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/credit-screening/crx.data', header=None)

fn = {'+': 1, '-': 0}

X = credit.replace('?',0).iloc[:,[1,2,7,10,14]]
y = credit.iloc[:,15].map(lambda x: fn.get(x,0))

1. Split the data into training and test set.
2. Construct a logistic regression model predicting y against X.
3. Score the model.
4. Construct a [confusion matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html) on the test data set and evaluate the model.
5. Analyse your results from Steps 3 and 4.

**1. Split the data into training and test set.**

I splitted data with function of sklearn.

In [323]:
from sklearn.model_selection import train_test_split

In [324]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

**2. Construct a logistic regression model predicting y against X.**

I assigned LogisticRegression. to `model`.  

In [325]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix

In [326]:
model = LogisticRegression(max_iter=1500)
model.fit(X_train,y_train)

LogisticRegression(max_iter=1500)

**3. Score the model.**

Scoring model with **model.score**

In [327]:
print(model.score(X_test,y_test))

0.7368421052631579


**4. Construct a [confusion matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html) on the test data set and evaluate the model.**

I constructed it with the function `confusion_matrix`

In [328]:
y_predict = model.predict(X_test)
confusion_matrix(y_test,y_predict)

array([[108,  17],
       [ 43,  60]], dtype=int64)

**5. Analyse your results from Steps 3 and 4.**

From step 3 we see that our score is 0.73 which means model predicted 73 percent of the test set true and the rest false. In Step 4, we see the distribution of these predictions. Confusion matrix shows that 43 guesses made wrong in guessing y which records only 1 or 0. It means 43 times model guessed 1 while the true value was 0. Out of 151 0 values, 43 of them was wrong. 108 / 151 = 0.715 which might not be very reliable model.  